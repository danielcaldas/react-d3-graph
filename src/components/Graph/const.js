export default {
    COORDS_SEPARATOR: ',',
    FORCE_IDEAL_STRENGTH: -100,
    FORCE_X: 0.06,
    FORCE_Y: 0.06,
    GRAPH_CONTAINER_ID: 'graph-container-zoomable',
    GRAPH_WRAPPER_ID: 'graph-wrapper',
    KEYWORDS: {
        SAME: 'SAME'
    },
    LINK_CLASS_NAME: 'link',
    LINK_IDEAL_DISTANCE: 100, // @todo: Expose as configurable,
    NODE_CLASS_NAME: 'node',
    SYMBOLS: {
        CIRCLE: 'circle',
        CROSS: 'cross',
        DIAMOND: 'diamond',
        SQUARE: 'square',
        STAR: 'star',
        TRIANGLE: 'triangle',
        WYE: 'wye'
    }
};
