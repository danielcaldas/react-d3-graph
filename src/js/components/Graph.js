import React from 'react';
import * as d3 from 'd3';
import { has as _has, merge as _merge } from 'lodash';

import CONST from './const';
import DEFAULT_CONFIG from './graph.config';

export default class Graph extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        let graph = _has(this, 'props.data') && this.props.data || {};
        let config = DEFAULT_CONFIG;
        if (_has(this, 'props.config')) {
            config = _merge(config, this.props.config);
        }

        /*----------------------------------------
            Some helper functions
        ----------------------------------------*/
        function isNumber(n) {
            return !isNaN(parseFloat(n)) && isFinite(n);
        }

        function tick() {
            node.attr('transform', function(d) {
                return 'translate(' + d.x + ',' + d.y + ')';
            });
            text.attr('transform', function(d) {
                return 'translate(' + d.x + ',' + d.y + ')';
            });

            link.attr('x1', function(d) {
                return d.source.x;
            }).attr('y1', function(d) {
                return d.source.y;
            }).attr('x2', function(d) {
                return d.target.x;
            }).attr('y2', function(d) {
                return d.target.y;
            });

            node.attr('cx', function(d) {
                return d.x;
            }).attr('cy', function(d) {
                return d.y;
            });
        }

        /*----------------------------------------
            Drag & Drop
         ----------------------------------------*/
        function dragstart(d) {
            if (!d3.event.active) simulation.alphaTarget(0.3).restart();
            d.fx = d.x;
            d.fy = d.y;
        }

        function dragmove(d) {
            d.fx = d3.event.x;
            d.fy = d3.event.y;
        }

        function dragend(d) {
            // if (!d3.event.active) simulation.alphaTarget(0);
            // Set this to null and you will see dragged node trying to comeback to initial position
            if (!config.nodeFixedAfterDropped) {
                d.fx = null;
                d.fy = null;
            }
        }
        /*----------------------------------------*/

        let circle;
        let link;
        let linkedByIndex = {};
        let node;
        let text;
        let tocolor = 'fill';
        let towhite = 'stroke';

        let color = d3.scaleLinear().domain([config.minScore, (config.minScore + config.maxScore) / 2, config.maxscore]).range(['lime', 'yellow', 'red']);
        let size = d3.scalePow().exponent(1).domain([1, 100]).range([8, 24]);
        let svg = d3.select('body').append('svg');

        svg.style('cursor', 'move');
        svg.style('width', config.width);
        svg.style('height', config.height);
        svg.style('border', '1px solid black');
        let g = svg.append('g');

        if (_has(graph, 'links')) {
            for (let d of graph.links) {
                linkedByIndex[`${d.source},${d.target}`] = true
            }
        }

        function strokeStyle(d) {
            return isNumber(d.score) && d.score >= 0 ? color(d.score) : config.defaultLinkColor;
        }

        link = g.selectAll('.link')
                .data(graph.links)
                .enter()
                .append('line').attr('class', 'link').style('stroke-width', config.strokeThickness).style('stroke', strokeStyle);

        const customNodeDrag = d3.drag()
                                .on('start', dragstart)
                                .on('drag', dragmove)
                                .on('end', dragend);

        node = g.selectAll('.node').data(graph.nodes).enter().append('g').attr('class', 'node').call(customNodeDrag);

        if (config.outline) {
            tocolor = 'stroke';
            towhite = 'fill';
        }

        /**
         * Converts a string that specifies a symbol into a concrete instance
         * of d3 symbol.
         * {@link https://github.com/d3/d3-shape/blob/master/README.md#symbol}
         * @param  {string} [typeName=CONST.SYMBOLS.CIRCLE] - the string that specifies the symbol type.
         * @return {Object} concrete instance of d3 symbol.
         */
        function convertTypeToD3Symbol(typeName=CONST.SYMBOLS.DIAMOND) {
            switch (typeName) {
                case CONST.SYMBOLS.CIRCLE:
                return d3.symbolCircle;
                break;
                case CONST.SYMBOLS.CROSS:
                return d3.symbolCross;
                break;
                case CONST.SYMBOLS.DIAMOND:
                return d3.symbolDiamond;
                break;
                case CONST.SYMBOLS.SQUARE:
                return d3.symbolSquare;
                break;
                case CONST.SYMBOLS.STAR:
                return d3.symbolStar;
                break;
                case CONST.SYMBOLS.TRIANGLE:
                return d3.symbolTriangle;
                break;
                case CONST.SYMBOLS.WYE:
                return d3.symbolWye;
                break;
            }
        }

        function buildSvgSymbol(config) {
            return d3.symbol()
                     .type((d) => convertTypeToD3Symbol(d.type))
                     .size((d) => Math.PI * Math.pow(size(d.size) || config.defaultNodeSize, 2));
        }

        circle = node
            .append('path')
            .attr('d', buildSvgSymbol(config))
            .style(tocolor, (d) => {
                if (d && d.color) {
                    return d.color;
                }
                return isNumber(d.score) && d.score >= 0 ? color(d.score) : config.defaultNodeColor;
            })
            .style('stroke-width', config.strokeThickness)
            .style(towhite, 'white');

        text = g.selectAll('.text').data(graph.nodes).enter().append('text').attr('dy', '.35em').style('font-size', config.defaultTextSize + 'px');

        if (config.textCenter) {
            text.text((d) => d.id).style('text-anchor', 'middle');
        } else {
            text.attr('dx', (d) => size(d.size) || config.defaultNodeSize)
                .text((d) => d[config.labelProperty] ? '\u2002' + d[config.labelProperty] : '\u2002' + d.id);
        }

        const forceX = d3.forceX(config.width / 2).strength(.05);
        const forceY = d3.forceY(config.height / 2).strength(.05);

        const simulation = d3.forceSimulation();

        simulation.force('link', d3.forceLink().distance(() => CONST.LINK_IDEAL_DISTANCE))
        .force('charge', d3.forceManyBody().strength(CONST.FORCE_IDEAL_STRENGTH))
        .force('x', forceX)
        .force('y', forceY);

        simulation.nodes(graph.nodes).on('tick', tick);
        simulation.force('link').links(graph.links);

        return (
            <g/>
        );
    }
}
