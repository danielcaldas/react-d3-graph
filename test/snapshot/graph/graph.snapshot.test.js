import React from 'react';
import renderer from 'react-test-renderer';

import Graph from '../../../src/components/graph/Graph';
import graphMock from './graph.mock.js';

describe('Snapshot - Graph Component', () => {
    let that = {};

    that.nodeColor = 'red';
    that.highlightColor = 'blue';
    that.svgSize = 600;
    that.highlightOpacity = 0.1;
    that.config = {
        height: that.svgSize,
        width: that.svgSize,
        nodeHighlightBehavior: true,
        highlightOpacity: that.highlightOpacity,
        staticGraph: true,
        collapsible: false,
        node: {
            color: that.nodeColor,
            highlightColor: that.highlightColor,
            size: 100
        },
        link: {
            highlightColor: that.highlightColor
        }
    };
    that.mouseOverNodeCallback = jest.fn();

    beforeEach(() => {
        that.graph = renderer.create(
            <Graph id="graphId" data={graphMock} config={that.config} onMouseOverNode={that.mouseOverNodeCallback} />
        );

        that.tree = that.graph.toJSON();
    });

    test('should match snapshot', () => {
        expect(that.tree).toMatchSnapshot();
    });
});
