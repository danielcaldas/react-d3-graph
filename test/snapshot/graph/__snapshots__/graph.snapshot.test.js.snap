// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot - Graph Component should match snapshot 1`] = `
<div
  id="graphId-graph-wrapper"
>
  <svg
    style={
      Object {
        "height": 600,
        "width": 600,
      }
    }
  >
    <g
      id="graphId-graph-container-zoomable"
    >
      <path
        className="link"
        d="M40,200A0,0 0 0,1 140,20"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="40"
        x2="140"
        y1="200"
        y2="20"
      />
      <path
        className="link"
        d="M40,200A0,0 0 0,1 20,110"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="40"
        x2="20"
        y1="200"
        y2="110"
      />
      <path
        className="link"
        d="M30,184A0,0 0 0,1 305,745"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="30"
        x2="305"
        y1="184"
        y2="745"
      />
      <path
        className="link"
        d="M200,300A0,0 0 0,1 305,745"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="200"
        x2="305"
        y1="300"
        y2="745"
      />
      <path
        className="link"
        d="M120,270A0,0 0 0,1 305,745"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="120"
        x2="305"
        y1="270"
        y2="745"
      />
      <path
        className="link"
        d="M20,110A0,0 0 0,1 305,745"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="20"
        x2="305"
        y1="110"
        y2="745"
      />
      <path
        className="link"
        d="M190,609A0,0 0 0,1 305,745"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="190"
        x2="305"
        y1="609"
        y2="745"
      />
      <path
        className="link"
        d="M40,200A0,0 0 0,1 101,201"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="40"
        x2="101"
        y1="200"
        y2="201"
      />
      <path
        className="link"
        d="M40,200A0,0 0 0,1 2,200"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="40"
        x2="2"
        y1="200"
        y2="200"
      />
      <path
        className="link"
        d="M404,404A0,0 0 0,1 14,30"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="404"
        x2="14"
        y1="404"
        y2="30"
      />
      <path
        className="link"
        d="M305,745A0,0 0 0,1 20,110"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="305"
        x2="20"
        y1="745"
        y2="110"
      />
      <path
        className="link"
        d="M40,200A0,0 0 0,1 14,250"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="40"
        x2="14"
        y1="200"
        y2="250"
      />
      <path
        className="link"
        d="M30,184A0,0 0 0,1 190,609"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="30"
        x2="190"
        y1="184"
        y2="609"
      />
      <path
        className="link"
        d="M14,259A0,0 0 0,1 190,609"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="14"
        x2="190"
        y1="259"
        y2="609"
      />
      <path
        className="link"
        d="M409,500A0,0 0 0,1 190,609"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="409"
        x2="190"
        y1="500"
        y2="609"
      />
      <path
        className="link"
        d="M520,123A0,0 0 0,1 14,259"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="520"
        x2="14"
        y1="123"
        y2="259"
      />
      <path
        className="link"
        d="M190,609A0,0 0 0,1 14,259"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="190"
        x2="14"
        y1="609"
        y2="259"
      />
      <path
        className="link"
        d="M305,745A0,0 0 0,1 120,270"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="305"
        x2="120"
        y1="745"
        y2="270"
      />
      <path
        className="link"
        d="M14,20A0,0 0 0,1 120,270"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="14"
        x2="120"
        y1="20"
        y2="270"
      />
      <path
        className="link"
        d="M305,745A0,0 0 0,1 200,300"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="305"
        x2="200"
        y1="745"
        y2="300"
      />
      <path
        className="link"
        d="M40,200A0,0 0 0,1 200,300"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="40"
        x2="200"
        y1="200"
        y2="300"
      />
      <path
        className="link"
        d="M200,300A0,0 0 0,1 40,200"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="200"
        x2="40"
        y1="300"
        y2="200"
      />
      <path
        className="link"
        d="M20,1A0,0 0 0,1 90,90"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="20"
        x2="90"
        y1="1"
        y2="90"
      />
      <path
        className="link"
        d="M90,656A0,0 0 0,1 90,90"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="90"
        x2="90"
        y1="656"
        y2="90"
      />
      <path
        className="link"
        d="M190,609A0,0 0 0,1 30,184"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="190"
        x2="30"
        y1="609"
        y2="184"
      />
      <path
        className="link"
        d="M305,745A0,0 0 0,1 30,184"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="305"
        x2="30"
        y1="745"
        y2="184"
      />
      <path
        className="link"
        d="M190,609A0,0 0 0,1 409,500"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="190"
        x2="409"
        y1="609"
        y2="500"
      />
      <path
        className="link"
        d="M120,270A0,0 0 0,1 14,20"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "fill": "none",
            "opacity": 1,
            "stroke": "#d3d3d3",
            "strokeWidth": 1.5,
          }
        }
        x1="120"
        x2="14"
        y1="270"
        y2="20"
      />
      <g
        className="node"
        cx="40"
        cy="200"
        id="Harry"
        transform="translate(40,200)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Harry
        </text>
      </g>
      <g
        className="node"
        cx="140"
        cy="20"
        id="Sally"
        transform="translate(140,20)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Sally
        </text>
      </g>
      <g
        className="node"
        cx="20"
        cy="110"
        id="Mario"
        transform="translate(20,110)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Mario
        </text>
      </g>
      <g
        className="node"
        cx="30"
        cy="184"
        id="Sarah"
        transform="translate(30,184)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Sarah
        </text>
      </g>
      <g
        className="node"
        cx="305"
        cy="745"
        id="Alice"
        transform="translate(305,745)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Alice
        </text>
      </g>
      <g
        className="node"
        cx="200"
        cy="300"
        id="Eveie"
        transform="translate(200,300)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Eveie
        </text>
      </g>
      <g
        className="node"
        cx="120"
        cy="270"
        id="Peter"
        transform="translate(120,270)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Peter
        </text>
      </g>
      <g
        className="node"
        cx="190"
        cy="609"
        id="James"
        transform="translate(190,609)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          James
        </text>
      </g>
      <g
        className="node"
        cx="101"
        cy="201"
        id="Carol"
        transform="translate(101,201)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Carol
        </text>
      </g>
      <g
        className="node"
        cx="2"
        cy="200"
        id="Nicky"
        transform="translate(2,200)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Nicky
        </text>
      </g>
      <g
        className="node"
        cx="404"
        cy="404"
        id="Bobby"
        transform="translate(404,404)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Bobby
        </text>
      </g>
      <g
        className="node"
        cx="14"
        cy="30"
        id="Frank"
        transform="translate(14,30)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Frank
        </text>
      </g>
      <g
        className="node"
        cx="14"
        cy="250"
        id="Lynne"
        transform="translate(14,250)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Lynne
        </text>
      </g>
      <g
        className="node"
        cx="14"
        cy="259"
        id="Roger"
        transform="translate(14,259)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Roger
        </text>
      </g>
      <g
        className="node"
        cx="409"
        cy="500"
        id="Maddy"
        transform="translate(409,500)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Maddy
        </text>
      </g>
      <g
        className="node"
        cx="520"
        cy="123"
        id="Sonny"
        transform="translate(520,123)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Sonny
        </text>
      </g>
      <g
        className="node"
        cx="14"
        cy="20"
        id="Johan"
        transform="translate(14,20)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Johan
        </text>
      </g>
      <g
        className="node"
        cx="20"
        cy="1"
        id="Henry"
        transform="translate(20,1)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Henry
        </text>
      </g>
      <g
        className="node"
        cx="90"
        cy="90"
        id="Mikey"
        transform="translate(90,90)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Mikey
        </text>
      </g>
      <g
        className="node"
        cx="90"
        cy="656"
        id="Elric"
        transform="translate(90,656)"
      >
        <path
          cursor="pointer"
          d="M5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,-5.641895835477563,0A5.641895835477563,5.641895835477563,0,1,1,5.641895835477563,0"
          fill="red"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          opacity={1}
          stroke="none"
          strokeWidth={1.5}
        />
        <text
          dx={10.5}
          dy=".35em"
          fill="black"
          fontSize={8}
          fontWeight="normal"
          opacity={1}
        >
          Elric
        </text>
      </g>
    </g>
  </svg>
</div>
`;
